videostore/indexer.go

```diff
@@ -0,0 +1,451 @@
+	if err != nil {
+		ix.logger.Debugf("failed to get video info, unreadable file will not be indexed: %w", err)
+		return nil
+	}
+	durationMs := info.duration.Milliseconds()
+
+	query := fmt.Sprintf("INSERT INTO %s (file_path, start_time_unix, duration_ms, size_bytes) VALUES (?, ?, ?, ?);", segmentsTableName)
+	_, err = ix.db.Exec(query, filePath, startTime.Unix(), durationMs, fileSize)
+	if err != nil {
+		return fmt.Errorf("failed to insert segment into index: %w", err)
+	}
+
+	startTimeStr := startTime.UTC().Format(TimeFormat) + "Z"
+	ix.logger.Debugw("indexed new file", "file", filePath, "start_time", startTimeStr, "duration_ms", durationMs, "size_bytes", fileSize)
+	return nil
+}
+
+// cleanupDB deletes old records and returns the paths of the to-be-deleted segment files.
+func (ix *indexer) cleanupDB(ctx context.Context) ([]string, error) {
+	maxStorageSizeBytes := int64(ix.storageMaxGB) * gigabyte
+	currentSizeBytes, err := getDirectorySize(ix.storagePath)
+	if err != nil {
+		return nil, err
+	}
+	if currentSizeBytes < maxStorageSizeBytes {
+		return nil, nil
+	}
+	bytesToDelete := currentSizeBytes - maxStorageSizeBytes
+
+	allSegments, err := ix.getSegmentsAscTime(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if len(allSegments) == 0 {
+		return nil, errors.New("cleanup required based on disk size, but indexer found no segments")
+	}
+
+	var (
+		pathsToDelete []string
+		bytesDeleted  int64
+	)
+	for _, segment := range allSegments {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+		pathsToDelete = append(pathsToDelete, segment.FilePath)
+		bytesDeleted += segment.SizeBytes
+		if bytesDeleted >= bytesToDelete {
+			break
+		}
+	}
+	if len(pathsToDelete) == 0 {
+		return nil, nil
+	}
+	deletedFilePaths, err := ix.deleteSegmentsAndReturnPaths(ctx, pathsToDelete)
+	if err != nil {
+		return nil, err
+	}
+	ix.logger.Infof("deleted %d segments from db (soon to be deleted from disk): %v", len(deletedFilePaths), deletedFilePaths)
+	return deletedFilePaths, nil
+}
+
+// deleteSegmentsAndReturnPaths is a helper function that deletes segments from the DB
+// and returns the paths of the deleted segments.
+func (ix *indexer) deleteSegmentsAndReturnPaths(ctx context.Context, paths []string) ([]string, error) {
+	if !ix.setupDone {
+		return nil, errors.New("indexer setup not complete")
+	}
+	if len(paths) == 0 {
+		return nil, nil
+	}
+
+	// Build the SQL query
+	placeholders := make([]string, len(paths))
+	args := make([]interface{}, len(paths))
+	for i, path := range paths {
+		placeholders[i] = "?"
+		args[i] = path
+	}
+	//nolint:gosec // this is safe because segmentsTableName is a constant and placeholders are '?'
+	query := fmt.Sprintf("DELETE FROM %s WHERE file_path IN (%s) RETURNING file_path", segmentsTableName, strings.Join(placeholders, ", "))
+
+	// QueryContext is used because the DELETE statement includes a RETURNING clause to get the deleted file paths.
+	rows, err := ix.db.QueryContext(ctx, query, args...)
```

Comment:
why can't you pass in `paths...`?

Reply:
`cannot use paths (variable of type []string) as []any value in argument to ix.db.QueryContext`